crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -3567,14 +3608,40 @@ impl<TX: DbTxMut + DbTx> BlockWriter for DatabaseProvider<TX> {
             }
         }
 
-        let block_indices = StoredBlockBodyIndices { first_tx_num, tx_count };
-        self.tx.put::<tables::BlockBodyIndices>(block_number, block_indices.clone())?;
-        durations_recorder.record_relative(metrics::Action::InsertBlockBodyIndices);
+        Ok(())
+    }
 
-        if !block_indices.is_empty() {
-            self.tx.put::<tables::TransactionBlocks>(block_indices.last_tx_num(), block_number)?;
-            durations_recorder.record_relative(metrics::Action::InsertTransactionBlocks);
-        }
+    /// Inserts the block into the database, always modifying the following tables:
+    /// * [`CanonicalHeaders`](tables::CanonicalHeaders)
+    /// * [`Headers`](tables::Headers)
+    /// * [`HeaderNumbers`](tables::HeaderNumbers)
+    /// * [`HeaderTerminalDifficulties`](tables::HeaderTerminalDifficulties)
+    /// * [`BlockBodyIndices`](tables::BlockBodyIndices)
+    ///
+    /// If there are transactions in the block, the following tables will be modified:
+    /// * [`Transactions`](tables::Transactions)
+    /// * [`TransactionBlocks`](tables::TransactionBlocks)
+    ///
+    /// If ommers are not empty, this will modify [`BlockOmmers`](tables::BlockOmmers).
+    /// If withdrawals are not empty, this will modify
+    /// [`BlockWithdrawals`](tables::BlockWithdrawals).
+    /// If requests are not empty, this will modify [`BlockRequests`](tables::BlockRequests).
+    ///
+    /// If the provider has __not__ configured full sender pruning, this will modify
+    /// [`TransactionSenders`](tables::TransactionSenders).
+    ///
+    /// If the provider has __not__ configured full transaction lookup pruning, this will modify
+    /// [`TransactionHashNumbers`](tables::TransactionHashNumbers).
+    fn insert_block(
+        &self,
+        block: SealedBlockWithSenders,
+    ) -> ProviderResult<StoredBlockBodyIndices> {
+        let block_number = block.number;
+        let mut durations_recorder = metrics::DurationsRecorder::default();
+
+        let block_indices =
+            self.insert_block_header_and_transaction_data(block.clone(), &mut durations_recorder)?;
+        self.insert_block_additional_data(block, &mut durations_recorder)?;
```

Comment:
do we need to clone here? or can we make `insert_block_header_and_transaction_data` accept `&SealedBlockWithSenders`